docs(operator): migration guide from kgo 1.6.x to ko 2.0 (#2796)
Kong/developer.konghq.com: app/operator/migrate/migrate-1.6.x-2.0.0.md

---
title: "Migrating from {{site.operator_product_name}} 1.6.x to Kong Operator 2.0.0"

> description: "Complete migration guide from {{site.operator_product_name}} (KGO) 1.6.x to Kong Operator (KO) 2.0.0."

Reviewer: ```suggestion
description: "Learn how to migrate from {{site.operator_product_name}} (KGO) 1.6.x to Kong Operator (KO) 2.0.0."
```

content_type: reference
layout: reference
products:
  - operator
breadcrumbs:
  - /operator/
works_on:
  - on-prem
  - konnect
tags:
  - migration
  - upgrade

related_resources:
  - text: "KonnectExtension Migration Prerequisites"
    url: /operator/migrate/konnectextension-konnectid-to-konnectgatewaycontrolplane/
  - text: "Kong Operator Changelog"
    url: /operator/changelog/
  - text: "Version Compatibility"
    url: /operator/reference/version-compatibility/

---

Kong Operator (KO) 2.0.0 represents a major version upgrade from {{site.operator_product_name}} (KGO) 1.6.x, introducing significant architectural improvements, enhanced Kubernetes-native features, and breaking changes that require careful migration planning.

## Prerequisites

> Before starting the migration, ensure you have:

Reviewer: ```suggestion
Before starting the migration, you need access to the Kubernetes cluster with admin privileges. Then, do the following:
```

> 1. **Backup your current configuration**
> 
> 2. **Verify current operator version**:
>    ```bash
>    helm list -n kong-system
>    ```

Reviewer: ````suggestion
1. Backup your current configuration.

2. Verify the current {{site.operator_product_name}} version:
   ```bash
   helm list -n kong-system
   ```
````

> 3. **Check for deprecated konnectID field** (if using KonnectExtensions):

Reviewer: ```suggestion
3. If you're using `KonnectExtensions`, check for the deprecated `konnectID` field:
```

   ```bash
   kubectl get konnectextensions --all-namespaces -o yaml | grep -A 5 -B 5 konnectID
   ```
   
   {:.warning}

>    > **Important**: If this command returns any results, you **must** migrate these KonnectExtensions before upgrading. Follow the [KonnectExtension Migration Guide](/operator/migrate/konnectextension-konnectid-to-konnectgatewaycontrolplane/) first.

Reviewer: ```suggestion
   > **Important**: If this command returns any results, you **must** migrate these `KonnectExtensions` before upgrading. Follow the [KonnectExtension Migration Guide](/operator/migrate/konnectextension-konnectid-to-konnectgatewaycontrolplane/) first.
```

> 4. **Access to Kubernetes cluster** with admin privileges

Reviewer: ```suggestion
```
Moving up in this section

> 5. **Verify cert-manager is installed** (required for conversion webhooks):

Reviewer: ```suggestion
5. Verify cert-manager is installed (required for conversion webhooks):
```

   
   {:.info}

>    > **Note**: Kong Operator 2.0.0 uses conversion webhooks that require TLS certificates managed by cert-manager. If cert-manager is not installed, follow the [cert-manager installation guide](https://cert-manager.io/docs/installation/) before proceeding. 

Author: > This is not required anymore. By default we use helm generated certs.

Helm-generated certificates are fine for development and testing, but for production environments, it’s recommended to use cert-manager. Unless you have a specific reason, I wouldn’t advise dropping cert-manager and relying solely on Helm-generated certs in the migration guide.

## Upgrade Kong Operator to 2.0.0

> The upgrade process requires several manual steps due to breaking changes in certificate management and CRD structure. Follow these steps carefully:

Reviewer: ```suggestion
The upgrade process requires several manual steps due to breaking changes in certificate management and CRD structure. Follow these steps carefully.
```

> ### Step 1: Uninstall the Existing KGO Release

Reviewer: ```suggestion
### Step 1: Uninstall the existing KGO release
```

First, uninstall the current {{site.operator_product_name}} release:

```bash
helm uninstall kgo -n kong-system
```

> ### Step 2: Label the Kong CA Certificate Secret

Reviewer: ```suggestion
### Step 2: Label the Kong CA certificate secret
```

> To ensure the Kong operator can properly reconcile the CA certificate secret, it must have the following label:

Reviewer: ```suggestion
To ensure the Kong Operator can properly reconcile the CA certificate secret, it must have the following label:
```

```bash
kubectl label secrets -n kong-system kong-operator-ca konghq.com/secret=true
```

### Step 3: Install Kong Operator 2.0.0

Install the new Kong Operator using Helm:

```bash
helm repo update kong
helm upgrade --install ko kong/kong-operator 
  -n kong-system \
  --create-namespace \
  --take-ownership \
  --set env.ENABLE_CONTROLLER_KONNECT=true \
  --set ko-crds.enabled=true \
  --set global.conversionWebhook.enabled=true \
  --set global.conversionWebhook.certManager.enabled=true 

> ```

Reviewer: I got this error when I ran this:

```
Error: Unable to continue with install: CustomResourceDefinition "aigateways.gateway-operator.konghq.com" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key "app.kubernetes.io/managed-by": must be set to "Helm"; annotation validation error: missing key "meta.helm.sh/release-name": must be set to "ko"; annotation validation error: missing key "meta.helm.sh/release-namespace": must be set to "default"
zsh: command not found: -n
```

Reviewer: Removing `--take-ownership` didn't fix it.

Author: The updated chart has not been released yet, and the current chart is not compatible with the latest KO version. I will notify you once the new chart becomes available.

> {:.info}

Reviewer: ```suggestion

{:.info}
```
This should get it to render correctly

> **Note**: The `--take-ownership` flag is required if CRDs or other resources were previously installed or managed by another tool (such as kubectl or a previous Helm release). This ensures Helm can properly manage and upgrade those resources as part of the new release.

> ### Step 5: Verify the Installation

Reviewer: ```suggestion
### Step 5: Verify the installation
```

> Verify that Kong Operator 2.0.0 is running correctly:
> 
> ```bash
> # Check the operator deployment
> kubectl get pod -n kong-system
> 
> # Check operator logs
>  kubectl logs -n kong-system -l app=ko-kong-operator
> ```

Reviewer: ````suggestion
Verify that Kong Operator 2.0.0 is running correctly.

1. Check the Kong Operator deployment:
   ```bash
   kubectl get pod -n kong-system
   ```

1. Check Kong Operator logs:
    ```sh
    kubectl logs -n kong-system -l app=ko-kong-operator
    ```
````
